mitiq/pec/channels.py

```diff
@@ -0,0 +1,219 @@
+    full_circ += _max_ent_state_circuit(2 * num_qubits)
+    full_circ += circuit
+    return simulator.simulate(full_circ).final_density_matrix  # type: ignore
+
+
+def _operation_to_choi(operation_tree: OP_TREE) -> np.ndarray:
+    """Returns the density matrix of the Choi state associated to the
+    input operation tree (e.g. a single operation or a sequence of operations).
+
+    The density matrix completely characterizes the quantum channel induced by
+    the input operation tree (including the effect of noise if present).
+
+    Args:
+        operation_tree: Nested list of operations.
+    Returns:
+        The density matrix of the Choi state associated to the input circuit.
+    """
+    circuit = Circuit(operation_tree)
+    return _circuit_to_choi(circuit)
+
+
+def _tensor_product_kraus(
+    local_kraus: List[np.ndarray], num_channels: int
+) -> List[np.ndarray]:
+    """Given the a local channel (defined by its kraus operators),
+    this functions returns the Kraus operators corresponding to tensor
+    product of "num_channels" copies of the input channel.
+    """
+    tensored_kraus = []
+    for kraus_tuple in product(local_kraus, repeat=num_channels):
+        kraus_product = np.eye(1)
+        for kraus in kraus_tuple:
+            kraus_product = np.kron(kraus_product, kraus)
+        tensored_kraus.append(kraus_product)
+    return tensored_kraus
+
+
+def global_depolarizing_kraus(
+    noise_level: float, num_qubits: int
+) -> List[np.ndarray]:
+    """Returns the kraus operators of a global depolarizing channel
+    at a given noise level.
+    """
+    noisy_op = DepolarizingChannel(noise_level, num_qubits)
+    return list(channel(noisy_op))
+
+
+def local_depolarizing_kraus(
+    noise_level: float, num_qubits: int
+) -> List[np.ndarray]:
+    """Returns the Kraus operators of the tensor product of local
+    depolarizing channels acting on each qubit.
+    """
+    local_kraus = global_depolarizing_kraus(noise_level, num_qubits=1)
+    return _tensor_product_kraus(local_kraus, num_channels=num_qubits)
+
+
+def amplitude_damping_kraus(
+    noise_level: float, num_qubits: int
+) -> List[np.ndarray]:
+    """Returns the Kraus operators of an amplitude damping
+    channel at a given noise level. If num_qubits > 1, the Kraus operators
+    corresponding to tensor product of many single-qubit amplitude damping
+    channels are returned.
+    """
+    noisy_op = AmplitudeDampingChannel(noise_level)
+    local_kraus = list(channel(noisy_op))
+    return _tensor_product_kraus(local_kraus, num_channels=num_qubits)
+
+
+def matrix_to_vector(density_matrix: np.ndarray) -> np.array:
+    """Reshapes a density matrix into a vector, according to the rule:
+    |i><j| --> |i,j>.
+    """
+    return density_matrix.flatten()
+
+
+def vector_to_matrix(vector: np.ndarray) -> np.array:
+    """Reshapes a vector to a density matrix, according to the rule:
+    |i,j> --> |i><j|.
+    """
+    dim = int(np.round(np.sqrt(vector.size)))
```

Comment:
Might be good to check if `vector.size` is a square number

Reply:
This happens in two functions and I expect to be a recurrent situation in the future, so I made a new function `_safe_sqrt` for handling this situation.